### PR TITLE
More minor fixes for the Onyx bump in E/App

### PR DESCRIPTION
**Pull request by chrispader**
@mountiny @s77rt @ikevin127 

### Details

This PR is adding more fixes to Onyx which are needed to resolve issues in the Onyx bump PR in E/App.

Fixes:

- `useOnyx` will never resolve loading state if the value from cache is `undefined` (because of `deepEqual` in `getSnapshot`)
- Update tests to reflect the new behavior of Onyx setting the cache. When we call Onyx (setter) methods this will set the value in cache synchronously (before storage promises are resolved). Therefore, we must first check for `initialValue` being set, before we can call these methods.

### Related Issues

https://github.com/Expensify/App/pull/42772

### Author Checklist

- [x] I linked the correct issue in the `### Related Issues` section above
- [x] I wrote clear testing steps that cover the changes made in this PR
    - [x] I added steps for local testing in the `Tests` section
    - [x] I tested this PR with a [High Traffic account](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#high-traffic-accounts) against the staging or production API to ensure there are no regressions (e.g. long loading states that impact usability).
- [x] I included screenshots or videos for tests on [all platforms](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#make-sure-you-can-test-on-all-platforms)
- [x] I ran the tests on **all platforms** & verified they passed on:
    - [x] Android / native
    - [x] Android / Chrome
    - [x] iOS / native
    - [x] iOS / Safari
    - [x] MacOS / Chrome / Safari
    - [x] MacOS / Desktop
- [x] I verified there are no console errors (if there's a console error not related to the PR, report it or open an issue for it to be fixed)
- [x] I followed proper code patterns (see [Reviewing the code](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md#reviewing-the-code))
    - [x] I verified that any callback methods that were added or modified are named for what the method does and never what callback they handle (i.e. `toggleReport` and not `onIconClick`)
    - [x] I verified that the left part of a conditional rendering a React component is a boolean and NOT a string, e.g. `myBool && <MyComponent />`.
    - [x] I verified that comments were added to code that is not self explanatory
    - [x] I verified that any new or modified comments were clear, correct English, and explained "why" the code was doing something instead of only explaining "what" the code was doing.
    - [x] I verified proper file naming conventions were followed for any new files or renamed files. All non-platform specific files are named after what they export and are not named "index.js". All platform-specific files are named for the platform the code supports as outlined in the README.
    - [x] I verified the JSDocs style guidelines (in [`STYLE.md`](https://github.com/Expensify/App/blob/main/contributingGuides/STYLE.md#jsdocs)) were followed
- [x] If a new code pattern is added I verified it was agreed to be used by multiple Expensify engineers
- [x] I followed the guidelines as stated in the [Review Guidelines](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md)
- [x] I tested other components that can be impacted by my changes (i.e. if the PR modifies a shared library or component like `Avatar`, I verified the components using `Avatar` are working as expected)
- [x] I verified all code is DRY (the PR doesn't include any logic written more than once, with the exception of tests)
- [x] I verified any variables that can be defined as constants (ie. in CONST.js or at the top of the file that uses the constant) are defined as such
- [x] I verified that if a function's arguments changed that all usages have also been updated correctly
- [x] If a new component is created I verified that:
    - [x] A similar component doesn't exist in the codebase
    - [x] All props are defined accurately and each prop has a `/** comment above it */`
    - [x] The file is named correctly
    - [x] The component has a clear name that is non-ambiguous and the purpose of the component can be inferred from the name alone
    - [x] The only data being stored in the state is data necessary for rendering and nothing else
    - [x] If we are not using the full Onyx data that we loaded, I've added the proper selector in order to ensure the component only re-renders when the data it is using changes
    - [x] For Class Components, any internal methods passed to components event handlers are bound to `this` properly so there are no scoping issues (i.e. for `onClick={this.submit}` the method `this.submit` should be bound to `this` in the constructor)
    - [x] Any internal methods bound to `this` are necessary to be bound (i.e. avoid `this.submit = this.submit.bind(this);` if `this.submit` is never passed to a component event handler like `onClick`)
    - [x] All JSX used for rendering exists in the render method
    - [x] The component has the minimum amount of code necessary for its purpose, and it is broken down into smaller components in order to separate concerns and functions
- [x] If any new file was added I verified that:
    - [x] The file has a description of what it does and/or why is needed at the top of the file if the code is not self explanatory
- [x] If the PR modifies a generic component, I tested and verified that those changes do not break usages of that component in the rest of the App (i.e. if a shared library or component like `Avatar` is modified, I verified that `Avatar` is working as expected in all cases)
- [x] If the `main` branch was merged into this PR after a review, I tested again and verified the outcome was still expected according to the `Test` steps.
- [x] I have checked off every checkbox in the PR author checklist, including those that don't apply to this PR.

### Screenshots/Videos
<details>
<summary>Android: Native</summary>

</details>

<details>
<summary>Android: mWeb Chrome</summary>

</details>

<details>
<summary>iOS: Native</summary>

</details>

<details>
<summary>iOS: mWeb Safari</summary>

</details>

<details>
<summary>MacOS: Chrome / Safari</summary>

</details>

<details>
<summary>MacOS: Desktop</summary>

</details>

### CODE REVIEW

File: lib/useOnyx.ts [hunk 1]
```diff
@@ -133,24 +131,29 @@ function useOnyx<TKey extends OnyxKey, TReturnValue = OnyxValue<TKey>>(key: TKey
             newFetchStatus = 'loading';
         }
 
-        // If data is not present in cache (if it's `undefined`) and `initialValue` is set during the first connection,
+        // If `options.initWithStoredValues` is explicitly set to `false`, the fetch status should always be `loaded`.
+        if (options?.initWithStoredValues === false) {
```
Comment (fabioh8010) [inline lib/useOnyx.ts:135]: Is this only applicable during the first connection or all the time? If only the first time we should use `isFirstConnectionRef.current`.

```ts
        if (isFirstConnectionRef.current && options?.initWithStoredValues === false) {
```

File: lib/useOnyx.ts [hunk 2]
```diff
@@ -133,24 +131,29 @@ function useOnyx<TKey extends OnyxKey, TReturnValue = OnyxValue<TKey>>(key: TKey
...
+            newFetchStatus = 'loaded';
+        }
+
+        // If data is not present in cache and `initialValue` is set during the first connection,
         // we set the new value to `initialValue` and fetch status to `loaded` since we already have some data to return to the consumer.
-        if (isFirstConnectionRef.current && newValue === undefined && options?.initialValue !== undefined) {
-            newValue = options?.initialValue as CachedValue<TKey, TReturnValue>;
+        if (isFirstConnectionRef.current && !hasCacheForKey && options?.initialValue !== null && options?.initialValue !== undefined) {
```
Comment (fabioh8010) [inline lib/useOnyx.ts:141]:
```suggestion
        if (isFirstConnectionRef.current && !hasCacheForKey && options?.initialValue !== undefined) {
```

According to the types we shouldn't be able to pass `null` to `initialValue`

File: tests/unit/useOnyxTest.ts [hunk 3]
```diff
@@ -257,6 +255,8 @@ describe('useOnyx', () => {
             expect(result.current[0]).toEqual('initial value');
             expect(result.current[1].status).toEqual('loaded');
 
+            Onyx.set(ONYXKEYS.TEST_KEY, 'test_id_1');
```
Comment (fabioh8010) [inline tests/unit/useOnyxTest.ts:258]: Just curious: Why did you move the operation to here? Shouldn't we have tests for both cases?

File: tests/unit/useOnyxTest.ts [hunk 4]
```diff
@@ -335,6 +329,12 @@ describe('useOnyx', () => {
             expect(result.current[0]).toEqual('initial value');
             expect(result.current[1].status).toEqual('loaded');
 
+            Onyx.set(ONYXKEYS.TEST_KEY, 'test1');
```
Comment (fabioh8010) [inline tests/unit/useOnyxTest.ts:332]: Same: Why did you move the operation to here? Shouldn't we have tests for both cases?

File: lib/useOnyx.ts (hunk 1 above)
Comment (chrispader) [reply lib/useOnyx.ts:135]: we can actually remove this.

File: tests/unit/useOnyxTest.ts (hunk 3 above)
Comment (chrispader) [reply tests/unit/useOnyxTest.ts:258]: The moment we use `Onyx.set` or `Onyx.merge` (or any setter function) it will store the value in cache synchronously and update the `storageKeys` Set in `OnyxCache`.\

If we set the key before asserting for the `initialValue`, we will always get `undefined`, because there IS a value in cache already, even though the storage promise hasn't resolved yet.

File: tests/unit/useOnyxTest.ts (hunk 4 above)
Comment (chrispader) [reply tests/unit/useOnyxTest.ts:332]: Same as above

File: tests/unit/useOnyxTest.ts (hunk 3 above)
Comment (fabioh8010) [reply tests/unit/useOnyxTest.ts:258]: Thanks for the explanation!

**Review by fabioh8010 — APPROVED** [review]

**Review by mountiny — DISMISSED** [review]
Just a NAB though seems like we are good to go

**Review by mountiny — APPROVED** [review]
Merging this to get it incorporated to the onyx bump PR per discussion here https://expensify.slack.com/archives/C05R2V5GM1S/p1718119019498549?thread_ts=1717193187.147939&cid=C05R2V5GM1S
